[FIRRTL][Dedup] Update few error messages

prithayan opened the pull request:
Update few error messages when `Dedup` fails for must dedup modules.
Check for mismatch in number of ports, and add special message for `RefType` ports.

youngar commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -260,33 +266,55 @@ struct Equivalence {
+      if (portNames)
         if (auto portNameAttr = portNames[portNo].dyn_cast<StringAttr>())
           portName = portNameAttr.getValue();
+      if (existsVal.getType().isa<RefType>()) {
+        diag.attachNote(opExists->getLoc())
+            << " contains a datap source port named '" + portName +
+                   "' that only exists in one of the modules";
+        diag.attachNote(opDoesNotExist->getLoc())
+            << "second module to be deduped that does not have the data tap";
```
s/datap/datatap, also maybe we should add a test for this code path.

youngar reviewed (approved):
LGTM

seldridge commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -250,6 +250,12 @@ struct Equivalence {
                    bType.cast<BundleType>());
+    if (aType.isa<RefType>() && bType.isa<RefType>())
+      return (check(diag,
+                    message + ", mismatch in data taps with two modules marked "
+                              "with must dedup.",
```
Nit: the difference is narrowly in reference types and not "data taps". For this type of thing, you may consider making the message longer with a suggestion of what to fix. E.g.: `mismatch in reference types (did you use a Grand Central Tap or View in a different order?)`. Ideally, we can map the source of the creator of the data tap back to the cause here. Jack's work towards adding source locators to annotations may help bridge this gap until all this moves into Chisel proper.

prithayan replied:
Fixed the error message. Still not incorporated the annotation location info.

seldridge commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -250,6 +250,14 @@ struct Equivalence {
                    bType.cast<BundleType>());
+    if (aType.isa<RefType>() && bType.isa<RefType>())
+      return (check(diag,
+                    message + ", mismatch in RefType (can be due to difference "
```
Mega nit: can you make these more complete sentences as opposed to fragments?

Consider:

"has a RefType port with a different base type ($type) of a port in the same position on a different module and these modules were marked as "must dedup". (This may be due to Grand Central Taps or Views being different between the two modules.)"

The note could then be:

"module port '$name' has different base type ($type)"

prithayan commented:
Thanks @seldridge and @youngar for the review, addressed all the comments, updated the error messages and added lit tests for all the cases. Merging this now.
